plugins/usd/module/vtkF3DFaceVaryingPointDispatcher.cxx

```diff
@@ -0,0 +1,134 @@
+  vtkPolyData* output = vtkPolyData::GetData(outputVector->GetInformationObject(0));
+
+  // early exit if all interpolations are "vertex"
+  vtkPointData* originalPointData = input->GetPointData();
+
+  vtkIdType nbArrays = originalPointData->GetNumberOfArrays();
+
+  bool earlyExit = true;
+
+  for (vtkIdType i = 0; i < nbArrays; i++)
+  {
+    vtkDataArray* originalArray = originalPointData->GetArray(i);
+
+    vtkInformation* info = originalArray->GetInformation();
+    int interpType = info->Get(vtkF3DFaceVaryingPointDispatcher::INTERPOLATION_TYPE());
+
+    if (interpType != 0) // vertex
+    {
+      earlyExit = false;
+      break;
+    }
+  }
+
+  if (earlyExit)
+  {
+    // nothing to do, just return the input
+    output->ShallowCopy(input);
+    return 1;
+  }
+
+  vtkPoints* originalPoints = input->GetPoints();
+  vtkCellArray* originalFaces = input->GetPolys();
+
+  vtkNew<vtkPoints> newPoints;
+  vtkNew<vtkCellArray> newFaces;
+
+  vtkIdType nbCells = originalFaces->GetNumberOfCells();
+  vtkIdType nbConnectivity = originalFaces->GetNumberOfConnectivityIds();
+
+  // the number of output points is the number of total cells connectivity
+  newPoints->SetNumberOfPoints(nbConnectivity);
+
+  vtkPointData* newPointData = output->GetPointData();
+
+  newPointData->ShallowCopy(originalPointData);
+
+  // Use the interpolation type metadata to know if the array must be replaced
+  for (vtkIdType i = 0; i < nbArrays; i++)
+  {
+    vtkDataArray* originalArray = originalPointData->GetArray(i);
+
+    vtkInformation* info = originalArray->GetInformation();
+    int interpType = info->Get(vtkF3DFaceVaryingPointDispatcher::INTERPOLATION_TYPE());
+
+    if (interpType == 0) // vertex
+    {
+      auto newArray = vtkSmartPointer<vtkDataArray>::Take(
+        vtkDataArray::CreateDataArray(originalArray->GetDataType()));
+      newArray->SetNumberOfComponents(originalArray->GetNumberOfComponents());
+      newArray->SetNumberOfTuples(nbConnectivity);
+      newArray->SetName(originalArray->GetName());
+
+      newPointData->AddArray(newArray);
+    }
+  }
+
+  std::vector<vtkIdType> cell;
+  cell.resize(originalFaces->GetMaxCellSize());
+
+  vtkIdType currentVertexIndex = 0;
+
+  for (vtkIdType i = 0; i < nbCells; i++)
+  {
+    vtkIdType cellSize;
+    const vtkIdType* cellPoints;
+    originalFaces->GetCellAtId(i, cellSize, cellPoints);
+
+    for (vtkIdType j = 0; j < cellSize; j++)
+    {
+      cell[j] = currentVertexIndex + j;
+      newPoints->SetPoint(currentVertexIndex + j, originalPoints->GetPoint(cellPoints[j]));
+
+      for (vtkIdType k = 0; k < nbArrays; k++)
+      {
+        vtkDataArray* originalArray = originalPointData->GetArray(k);
+
+        vtkInformation* info = originalArray->GetInformation();
+        int interpType = info->Get(vtkF3DFaceVaryingPointDispatcher::INTERPOLATION_TYPE());
+
+        if (interpType == 0) // vertex
+        {
+          vtkDataArray* newArray = newPointData->GetArray(originalArray->GetName());
```

Comment:
I'm not sure newArray is the right name here ?

Reply:
I'll replace `originalXXX` by `inputXXX` and `newXXX` by `outputXXX`